### PR TITLE
storageccl: allow URL params to be env vars instead

### PR DESCRIPTION
**Pull request by madelynnblue**
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/13932)

### CODE REVIEW

File: pkg/ccl/sqlccl/backup_cloud_test.go [hunk 1]
```diff
@@ -126,8 +129,11 @@ func TestBackupRestoreAzure(t *testing.T) {
 	defer cleanupFn()
 	prefix := fmt.Sprintf("TestBackupRestoreAzure-%d", timeutil.Now().UnixNano())
 	uri := url.URL{Scheme: "azure", Host: bucket, Path: prefix}
+	// Test that we can pass keys as params or env vars.
+	if err := os.Setenv(storageccl.AzureAccountNameParam, accountName); err != nil {
```
Comment (tamird) [inline pkg/ccl/sqlccl/backup_cloud_test.go]: please remember to clean up the environment at the end of the test

**Comment by mjibson** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/ccl/sqlccl/backup_cloud_test.go, line 133 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13932#-Kearnj7XY2INW_9M5G0-r1-133:-Kebk3xMamYr0TxHpnfs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df1d29f86135c9e47617cd1ba6e4bba9e0d2c5b0/pkg/ccl/sqlccl/backup_cloud_test.go#L133)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

please remember to clean up the environment at the end of the test
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13932)*

File: pkg/ccl/storageccl/export_storage.go [hunk 2]
```diff
@@ -99,6 +100,14 @@ func ExportStorageConfFromURI(path string) (roachpb.ExportStorage, error) {
 	return conf, nil
 }
 
+func getParamOrEnv(u *url.URL, name string) string {
+	s := u.Query().Get(name)
+	if s != "" {
+		return s
+	}
+	return envutil.EnvOrDefaultString(name, "")
```
Comment (dt) [inline pkg/ccl/storageccl/export_storage.go]: IIRC, `envutil` wants all calls at init time, and I though it also wanted everything prefixed with COCKROACH_?

**Review by dt — COMMENTED** [review]
I'm a little hesitant to rely on executing node's env - could introduce some strange unpredictable/ non-deterministic errors for clients behind pools / balances, vs the query carrying everything required to execute it. That said, using instance auth seems like a likely enough use-case that maybe that ship has sailed.

**Comment by mjibson** [comment]
The GCS client already uses stuff from the instance and not the URL, so I don't think this is too much of a surprise. I suppose we do need to consider what happens when multiple users are on the same node. Hmm, maybe this env stuff isn't a good idea.

**Comment by mjibson** [comment]
After discussion we've decided to stick with URL parameters for now. If users have good reasons for needing (or we find some ourselves) env vars, we can add those. The current thinking is that making these deterministic (i.e., not reliant on settings of the gateway node they run on) will be better for initial testing and usage.

Other things discussed:
- Fallback to env if param not present.
- Have a special sentinel param value like `getFromEnv` or something that uses the env, so it's not a fallback, but must be specified.

These above two would use the env from the gateway node. One problem with using env vars is the binary must be restarted to change them, which can be annoying or problematic. Putting keys in the params directly has some security implications in logs or running query lists. We also considered:

- Use zone configs to store and distribute access keys. This has avoids the security implications and restart problems of the above, but may have other problems not realized yet.

@bdarnell thoughts on this choice to stick with URL params for now?

**Comment by bdarnell** [comment]
Yeah, env vars don't seem like the right way to do this. Slightly better would be to read from a config file on the node (at least that way it could be updated without restarting the servers, and we might be able to share config files used by other tools), but even then the fact that it could vary from one node to another would be ugly. 

I think we need to either pass it in with the query and limit the visibility of full queries in logs and other places (which we should probably do regardless), or store it somewhere similar to the zone configs (but probably not the zone configs themselves - it should be an explicit lookup instead of gossip. It should probably also be pluggable for people who want to get their storage authentication from an external system like Vault)

**Pull request closed without merging**